cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1720,7 +1720,27 @@ class S3FileSystem::Impl : public std::enable_shared_from_this<S3FileSystem::Imp
   // a non-empty "directory".  This is a Minio-specific quirk, but we need
   // to handle it for unit testing.
 
-  Status IsEmptyDirectory(const std::string& bucket, const std::string& key, bool* out) {
+  // If this method is called after HEAD on "bucket/key" already returned a 404,
+  // can pass the given outcome to spare a spurious HEAD call.
+  Result<bool> IsEmptyDirectory(
+      const std::string& bucket, const std::string& key,
+      const S3Model::HeadObjectOutcome* previous_outcome = nullptr) {
```

Comment:
I don't know if this is one we enforce or not but the style guide prefers `util::optional<S3Model::HeadObjectOutcome>` (well, technically std::optional) instead of `const *`

https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs

Reply:
> Generally, use std::optional to represent optional by-value inputs, and use a const pointer when the non-optional form would have used a reference. 

Here, the non-optional form would have used a reference (we don't want to copy the entire outcome object).

Reply:
Oops, you are correct.  Thanks!